### PR TITLE
chore: OpenAPI specifications for all `/api/v2/advanced-filters` endpoints

### LINKED ISSUE #13752: Add OpenAPI specs for `/api/v2/advanced-filters` REST API endpoints
List of API endpoints

- `/api/v2/advanced-filters/`
- `/api/v2/advanced-filters/csv`
- `/api/v2/advanced-filters/methods`

### PR DESCRIPTION
**Pull request by akolotov**
Closes #13752

This PR adds OpenAPI specifications for all `/api/v2/advanced-filters` endpoints using `OpenApiSpex`.

### Endpoints covered

| Endpoint | Response schema |
|---|---|
| `GET /api/v2/advanced-filters` | `AdvancedFilter.Response` (paginated, with `search_params` lookup) |
| `GET /api/v2/advanced-filters/csv` | streamed `application/csv` (sync) or `AdvancedFilter.CsvExportAccepted` (async, `202`) |
| `GET /api/v2/advanced-filters/methods` | `AdvancedFilter.Method` (list) |

### Schemas

- **`AdvancedFilter`** — single item from `/advanced-filters`, modeling the union of native value transfers, internal transactions and token transfers (`Explorer.Chain.AdvancedFilter`). Carries `type`/`status` discriminators, optional `from`/`to`/`created_contract`, native `value`, polymorphic `total` (`anyOf` `Total`/`TotalERC721`/`TotalERC1155`/`TotalERC7984`), optional `token`, `fee`, and the keyset-cursor fields.
- **`AdvancedFilter.SearchParams`** — auxiliary lookup echoed in the response: `methods` (selector → name) and `tokens` (address → `Token`) for the filters supplied in the request.
- **`AdvancedFilter.Response`** — paginated envelope: `items`, `search_params`, `next_page_params`.
- **`AdvancedFilter.Method`** — `{method_id, name}` pair; `method_id` constrained to `^0x[0-9a-f]{8}$`.
- **`AdvancedFilter.CsvExportAccepted`** — `202 Accepted` body for async CSV export, carrying the `request_id` (UUID) to poll via `/api/v2/csv-exports/{request_id}`.
- **`AdvancedFilter.CsvExportError`** — `409`/`500` JSON error body for the CSV endpoint. Uses an `error` key rather than the `message` key carried by the reusable `ErrorResponses.*` schemas — the CSV controller predates that convention.

### Shared parameters

`/advanced-filters` and `/advanced-filters/csv` share the same filter surface, so parameters are extracted into module attributes and reused by both `operation` specs:

- **`@advanced_filter_query_params`** — 13 filter parameters (`transaction_types`, `methods`, `age_from`/`age_to`, `from_/to_address_hashes_to_include`/`_exclude`, `address_relation`, `amount_from`/`amount_to`, `token_contract_address_hashes_to_include`/`_exclude`). Descriptions document the comma-separated format, allowed values, the 20-entry caps, and the `native` literal.
- **`@advanced_filter_keyset_params`** — 5 keyset cursors. The three sub-indices accept either a numeric string or the literal `""`/`"null"` (modeled as `oneOf`) to signal items that aren't internal transactions / token transfers / batch items.
- **`@items_count_param`** — JSON-only cursor, omitted from the CSV operation.

### Cross-cutting changes

- **Controller**: integrated `OpenApiSpex.Plug.CastAndValidate`; all action pattern matches updated from string keys to atom keys. `tags(["advanced-filters"])` added.
- **`list_methods` cleanup**: head changed to `def list_methods(conn, %{q: query}) when is_binary(query)`, so the empty/missing-`q` case naturally falls through to the default-list branch with no special casing.
- **Test fix**: the pre-existing filter-stack test used a stale `tx_types` key with lowercase `coin_transfer`. Updated to `transaction_types=COIN_TRANSFER,ERC-20` to match both the spec and the controller's documented allowed values.

### Test coverage

- New `describe "/advanced_filters/methods"` block covering the default (no-`q`) listing — asserts non-empty list and per-item shape.
- Regression test for `q=""` returning `[]` (guards the new binary-only head from falling into the lookup branch).

All responses are validated against the OpenAPI schemas via `json_response/2`.

## Checklist for your Pull Request (PR)

- [x] I verified this PR does not break any public APIs, contracts, or interfaces that external consumers depend on.
- [x] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [ ] I updated documentation if needed:
  - [ ] General docs: submitted PR to [docs repository](https://github.com/blockscout/docs).
  - [ ] ENV vars: updated [env vars list](https://github.com/blockscout/docs/tree/main/setup/env-variables) and set version parameter to `master`.
  - [ ] Deprecated vars: added to [deprecated env vars list](https://github.com/blockscout/docs/tree/main/setup/env-variables/deprecated-env-variables).
- [x] If I modified API endpoints, I updated the Swagger/OpenAPI schemas accordingly and checked that schemas are asserted in tests, and highlighted the change in the PR description.
- [ ] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [ ] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

## Summary by CodeRabbit

* **New Features**
  * Added OpenAPI-powered API specs and formal schemas for Advanced Filters, methods, and CSV export responses.

* **Documentation**
  * Documented rich query parameters, responses (including async CSV export outcomes), and methods lookup.

* **Bug Fixes**
  * Improved request parameter parsing and pagination/cursor handling; stricter request/response validation.

* **Tests**
  * Added/updated tests for methods lookup, parameter mapping, cursor literal handling, and empty-query behavior.

### CODE REVIEW

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex [hunk 1]
```diff
@@ -55,6 +63,210 @@ defmodule BlockScoutWeb.API.V2.AdvancedFilterController do
...
+        oneOf: [
+          %OpenApiSpex.Schema{type: :string, pattern: ~r/^-?(?:[1-9][0-9]*|0)$/},
+          %OpenApiSpex.Schema{type: :string, enum: ["", "null"]}
+        ]
+      },
+      required: false,
+      description:
+        "Keyset cursor: index within an ERC-1155 batch token transfer. " <>
+          "Use an empty string or the literal `null` when the previous item was not part of a batch."
+    }
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex:231]: Agree — the response fields are non-negative, so negative cursor values are meaningless. Fixed: dropped the `-?` from all five keyset parameter patterns.

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex [hunk 2]
```diff
@@ -90,6 +302,25 @@ defmodule BlockScoutWeb.API.V2.AdvancedFilterController do
...
+    description:
+      "Streams the items matching the advanced filter criteria as a CSV file. " <>
+        "When asynchronous CSV export is enabled on the deployment, returns `202 Accepted` with a `request_id` " <>
+        "that can be polled via `/api/v2/csv-exports/{request_id}`; otherwise the CSV body is streamed inline.",
+    parameters: base_params() ++ @advanced_filter_query_params ++ @advanced_filter_keyset_params,
+    responses: [
+      ok: {"CSV file (sync export).", "application/csv", nil},
+      accepted:
+        {"Async export queued; poll `/api/v2/csv-exports/{request_id}` with the returned `request_id`.",
+         "application/json", Schemas.AdvancedFilter.CsvExportAccepted},
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex:304]: Won't fix. The entire project uses `application/csv` consistently — `csv_export_controller.ex` declares it in all five CSV endpoints and `put_resp_params/1` sets it as the runtime content type. Changing just this endpoint to `text/csv` would be inconsistent. If this should change, it needs to be a project-wide effort, not scoped to this PR.

File: apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+        nullable: true,
+        example: %{
+          "block_number" => 23_532_302,
+          "transaction_index" => 1,
+          "internal_transaction_index" => nil,
+          "token_transfer_index" => 0,
+          "token_transfer_batch_index" => nil,
+          "items_count" => 50
+        }
+      }
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex:218]: Won't fix. The centralized `General.paginated_response/1` helper defines `next_page_params` as an open object with only an example — every paginated endpoint in the project follows this pattern. This response schema can't use `paginated_response/1` directly (because of the extra `search_params` field), but it intentionally mirrors the same `next_page_params` structure for consistency.

Locking down `next_page_params` with explicit properties and `additionalProperties: false` on just this one endpoint would diverge from the project-wide convention and create a fragile special case. If strict typing for `next_page_params` is desired, the right approach is to update the centralized helper and apply it uniformly.

File: apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+    title: "AdvancedFilterItem",
+    type: :object,
+    properties: %{
+      hash: General.FullHash,
+      type: %Schema{
+        type: :string,
+        description:
+          "Kind of activity represented by the item. Values include `coin_transfer` and `contract_call` for " <>
+            "top-level transactions and internal transactions, as well as token-transfer type labels (e.g. `ERC-20`, " <>
+            "`ERC-721`, `ERC-1155`, `ERC-404`, `ERC-7984`) for token transfers.",
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex]: Good catch — the description incorrectly said `contract_call` instead of `contract_interaction` and omitted `contract_creation`. Fixed to list all three actual values from `assign_type/1`: `coin_transfer`, `contract_interaction`, `contract_creation`.

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex [hunk 5]
```diff
@@ -55,6 +63,210 @@ defmodule BlockScoutWeb.API.V2.AdvancedFilterController do
...
+    },
+    %OpenApiSpex.Parameter{
+      name: :token_contract_address_hashes_to_include,
+      in: :query,
+      schema: %OpenApiSpex.Schema{type: :string, nullable: true},
+      required: false,
+      description:
+        "Comma-separated list of token contract address hashes to include. Use the literal `native` to also " <>
+          "include native coin transfers. At most 20 unique entries (across include + exclude) are honored.",
+      example: "native,0xdac17f958d2ee523a2206206994597c13d831ec7"
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex:174]: Good catch. The code in `extract_filters/1` applies `Enum.take(@tokens_filter_limit)` to include and exclude lists independently, so each is capped to 20 (up to 40 total). Fixed the description to say "Each list (include and exclude) is capped to 20 entries separately."

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex [hunk 6]
```diff
@@ -90,6 +302,25 @@ defmodule BlockScoutWeb.API.V2.AdvancedFilterController do
...
+  operation :list_csv,
+    summary: "Export advanced-filter results as CSV",
+    description:
+      "Streams the items matching the advanced filter criteria as a CSV file. " <>
+        "When asynchronous CSV export is enabled on the deployment, returns `202 Accepted` with a `request_id` " <>
+        "that can be polled via `/api/v2/csv-exports/{request_id}`; otherwise the CSV body is streamed inline.",
+    parameters: base_params() ++ @advanced_filter_query_params ++ @advanced_filter_keyset_params,
+    responses: [
+      ok: {"CSV file (sync export).", "application/csv", nil},
+      accepted:
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex:302]: The code and spec are consistent — both use `application/csv`, which is the project-wide convention (all CSV endpoints in `csv_export_controller.ex` and `put_resp_params/1` use it). Updated the PR description to say `application/csv` instead of `text/csv`.

File: apps/block_scout_web/test/block_scout_web/controllers/api/v2/advanced_filter_controller_test.exs [hunk 7]
```diff
@@ -1389,13 +1389,34 @@ defmodule BlockScoutWeb.API.V2.AdvancedFilterControllerTest do
...
+    test "returns default list of methods", %{conn: conn} do
+      request = get(conn, "/api/v2/advanced-filters/methods")
+      assert response = json_response(request, 200)
+      assert is_list(response)
+      assert length(response) > 0
+
+      Enum.each(response, fn method ->
+        assert %{"method_id" => method_id, "name" => name} = method
+        assert method_id =~ ~r/^0x[0-9a-f]{8}$/
+        assert is_binary(name) and name != ""
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/test/block_scout_web/controllers/api/v2/advanced_filter_controller_test.exs:1419]: Won't change. This test specifically validates the **default methods list** — the hardcoded `@methods` module attribute containing well-known method signatures (`transfer`, `approve`, etc.). Those always have names. The `name != ""` assertion is intentional: it verifies the quality of the built-in data, not the general API contract.

The schema correctly allows empty strings for the broader case (runtime lookups where a name can't be resolved), but the default list should never contain empty names. These are two different concerns.

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex:174]: Addressed by 09aa966

File: apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex (hunk 4 above)
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex]: Addressed by 09aa966

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex:174]: Double check that it is addressed by 09aa966

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex [hunk 8]
```diff
@@ -55,6 +63,210 @@ defmodule BlockScoutWeb.API.V2.AdvancedFilterController do
...
+          "include native coin transfers. Each list (include and exclude) is capped to 20 entries separately.",
+      example: "native,0xdac17f958d2ee523a2206206994597c13d831ec7"
+    },
+    %OpenApiSpex.Parameter{
+      name: :token_contract_address_hashes_to_exclude,
+      in: :query,
+      schema: %OpenApiSpex.Schema{type: :string, nullable: true},
+      required: false,
+      description: "Comma-separated list of token contract address hashes to exclude.",
+      example: "0x0000000000000000000000000000000000000000"
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex:184]: Good catch. `prepare_token_address_hash/1` accepts `"native"` on both sides identically. Added the `native` literal documentation and the per-list cap note to the exclude parameter description as well.

File: apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+    ],
+    additionalProperties: false
+  })
+end
+
+defmodule BlockScoutWeb.Schemas.API.V2.AdvancedFilter.SearchParams do
+  @moduledoc """
+  Auxiliary lookup map returned alongside advanced-filter items that echoes
+  the resolved human-readable names of the `methods` and `tokens` used in the
+  request filters.
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex:161]: Agree that the schema over-promised. The existing code behavior is intentional and unchanged by this PR: `extract_filters/1` caps each list to 20 separately (for filtering), while `contract_address_hash_to_token_from_params/1` merges both and caps the metadata lookup to 20 combined (for the response echo). The fix is in the schema description, not the code — updated `search_params.tokens` to say "At most 20 entries are returned (combined across both lists)."

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex [hunk 10]
```diff
@@ -55,6 +63,212 @@ defmodule BlockScoutWeb.API.V2.AdvancedFilterController do
...
+    },
+    %OpenApiSpex.Parameter{
+      name: :methods,
+      in: :query,
+      schema: %OpenApiSpex.Schema{type: :string, nullable: true},
+      required: false,
+      description:
+        "Comma-separated list of 4-byte contract method selectors (lowercase, `0x`-prefixed). At most 20 unique " <>
+          "entries are honored; invalid entries are dropped.",
+      example: "0xa9059cbb,0x095ea7b3"
```
[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex:89]: > The current description makes the API contract stricter than the actual behavior.

Do you see this as a problem? The lowercase is preferable. Event if the code accepts the uppercase there is no need to document this.

[the remark this answers is not in the document]
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex:89]: Add this to your learning

File: apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+            "for token transfers.",
+        nullable: false
+      },
+      status: %Schema{
+        type: :string,
+        description:
+          "Execution status of the parent transaction. One of `pending`, `awaiting_internal_transactions`, " <>
+            "`success`, or a free-form error reason string when the transaction reverted (e.g. `Reverted`).",
+        nullable: false
+      },
```
Comment (vbaranov) [inline apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex:47]: What’s the reasoning behind not using enums for the `type` and `status` properties?

File: apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+        nullable: true,
+        description: "Address of the contract deployed by this item. `null` unless the item is a contract creation."
+      },
+      value: %Schema{
+        type: :string,
+        pattern: General.integer_pattern(),
+        nullable: true,
+        description:
+          "Native coin amount transferred, in the chain's base unit (e.g. wei). `null` for token-transfer items."
+      },
```
Comment (vbaranov) [inline apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex:70]: Worth adding `minimum: 0` for consistency?

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex [hunk 13]
```diff
@@ -55,6 +63,212 @@ defmodule BlockScoutWeb.API.V2.AdvancedFilterController do
...
+          "exclude native coin transfers. Each list (include and exclude) is capped to 20 entries separately.",
+      example: "0x0000000000000000000000000000000000000000"
+    }
+  ]
+
+  @advanced_filter_keyset_params [
+    %OpenApiSpex.Parameter{
+      name: :block_number,
+      in: :query,
+      schema: %OpenApiSpex.Schema{type: :string, pattern: ~r/^(?:[1-9][0-9]*|0)$/},
```
Comment (vbaranov) [inline apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex]: effectively, `General.integer_pattern()` and 4 times below.

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex [hunk 14]
```diff
@@ -55,6 +63,212 @@ defmodule BlockScoutWeb.API.V2.AdvancedFilterController do
...
+    },
+    %OpenApiSpex.Parameter{
+      name: :internal_transaction_index,
+      in: :query,
+      schema: %OpenApiSpex.Schema{
+        oneOf: [
+          %OpenApiSpex.Schema{type: :string, pattern: ~r/^(?:[1-9][0-9]*|0)$/},
+          %OpenApiSpex.Schema{type: :string, enum: ["", "null"]}
+        ]
+      },
```
Comment (vbaranov) [inline apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex]: Pattern is repeated 3 times in the given module. Maybe worth to extract to `apps/block_scout_web/lib/block_scout_web/schemas/api/v2/general`?

File: apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+      token: %Schema{
+        allOf: [Token],
+        nullable: true,
+        description: "Token contract metadata. Populated only for token-transfer items; `null` otherwise."
+      },
+      timestamp: %Schema{
+        type: :string,
+        format: :"date-time",
+        nullable: false,
+        description: "Block timestamp of the parent transaction."
```
Comment (vbaranov) [inline apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex]: effectively, `BlockScoutWeb.Schemas.API.V2.General.Timestamp`

**Review by vbaranov — COMMENTED** [review]
Let's list "advanced-filters" category last in the `@default_api_categories` in `BlockScoutWeb.Specs.Public` module. This ensure "legacy" category will be the last in the swagger file.

**Comment by akolotov** [comment]
@vbaranov 

> Let's list "advanced-filters" category last in the `@default_api_categories` in `BlockScoutWeb.Specs.Public` module. This ensure "legacy" category will be the last in the swagger file.

Done in 59e5284 — `"advanced-filters"` is now appended to `@default_api_categories` (last, after `"withdrawals"`).

File: apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex (hunk 11 above)
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex:47]: Different reasoning for each:

**`type`**: agreed, switched to `enum` in 59e5284. The set is closed: `coin_transfer` / `contract_interaction` / `contract_creation` come from hard-coded branches in `Explorer.Chain.AdvancedFilter.assign_type/1`, and the `ERC-*` values come from `Token.valid_types/0`. Sync comment added pointing at both sources.

**`status`**: kept as `type: :string`. The set is open-ended — `BlockScoutWeb.API.V2.TransactionView.format_status/1` returns the well-known atoms `:pending` / `:awaiting_internal_transactions` / `:success` for the happy path, but for `{:error, reason}` it returns the **free-form revert reason string** from the contract (e.g. `"Reverted"`, custom revert messages — anything the contract emits). An `enum` would misrepresent the API as closed; modeling it as `oneOf [enum, string]` adds no validation over plain `string`. The well-known values are listed in the description for clients.

File: apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex (hunk 12 above)
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex:70]: Addressed in 59e5284, with a small structural nuance: `value` (and `fee`, which has the same shape) is `type: :string` because wei amounts overflow JS-safe integers. JSON Schema's `minimum:` keyword is numeric-only and is ignored on `type: :string`, so the literal suggestion can't be applied — but the intent is right (these amounts are non-negative on chain).

Introduced `General.non_negative_integer_pattern/0` in `general.ex` (mirrors `@integer_pattern` minus the `-?`) and switched `value` and `fee` to use it. Same helper is reused for the keyset cursors per your other comments.

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex (hunk 13 above)
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex]: Deduplicated in 59e5284, but pointed at a new helper rather than `General.integer_pattern()`. The keyset cursor pattern (`~r/^(?:[1-9][0-9]*|0)$/`) is strictly non-negative; `General.integer_pattern()` is `~r"^-?([1-9][0-9]*|0)$"` and would silently loosen the validation to accept `-1`/etc.

Added `General.non_negative_integer_pattern/0` and used it across all 5 sites (the two `block_number`/`transaction_index` keyset params here, plus the inner branches of the three `oneOf` schemas — see the next comment).

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex (hunk 14 above)
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/advanced_filter_controller.ex]: Done in 59e5284. Extracted into a new leaf schema:
`BlockScoutWeb.Schemas.API.V2.General.IntegerStringOrEmptyOrNullLiteral`
at `schemas/api/v2/general/integer_string_or_empty_or_null_literal.ex`. The three keyset params (`internal_transaction_index`, `token_transfer_index`, `token_transfer_batch_index`) now reference it directly. Per-parameter `description:` stays on the `%Parameter{}` struct, so the leaf schema is description-less, matching neighbors like `IntegerString` and `Timestamp`.

Also added a regression test (38th test in the file) that sends the literal `"null"` for all three cursors and asserts 200 — the round-trip pagination tests already exercise the empty-string branch implicitly via Phoenix's `nil`-to-empty encoding, but the literal `"null"` branch had no coverage.

File: apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex (hunk 15 above)
Comment (akolotov) [reply apps/block_scout_web/lib/block_scout_web/schemas/api/v2/advanced_filter.ex]: Reused in 59e5284, but with a small twist worth flagging.

The straightforward `timestamp: General.Timestamp` (the form used by all 14 other call sites in the codebase) drops the per-property description, which felt unfortunate — a downstream reader of the spec for this endpoint loses the context that the field represents the parent transaction's block timestamp specifically. The obvious workaround `%Schema{allOf: [General.Timestamp], description: "..."}` actually crashes the OpenApiSpex test cast layer with `Protocol.UndefinedError: protocol Enumerable not implemented for DateTime` — `Cast.AllOf.cast_all_of/2` runs the `format: :"date-time"` cast (string → `%DateTime{}`) and then enumerates the result as a map. Works for object schemas like `allOf: [Address]`, breaks for primitives whose format casts to non-primitive structs.

Settled on `Helper.extend_schema(General.Timestamp.schema(), description: "...")` — uses the existing helper (which explicitly supports the `:description` overlay), keeps the timestamp definition single-sourced via `General.Timestamp.schema()`, and preserves the per-property description. Resolved spec output now carries both `format: date-time, nullable: false` from the leaf schema *and* the description, with `x-struct: …General.Timestamp` showing the source.

Worth flagging as a potential follow-up cleanup: the other 14 `General.Timestamp` consumers also lack per-property descriptions. Either adding a generic description to `General.Timestamp` itself, or applying the same `Helper.extend_schema` overlay at each site, would help downstream readers (humans and AI tooling) understand what each timestamp represents — happy to file a separate PR if desired.

**Review by vbaranov — APPROVED** [review]
